pkg/storage/crd/storage_interface.go

```diff
@@ -0,0 +1,547 @@
+	if t.hardDelete {
+		// if we are hard-deleting this item, then propagate this delete to the core API server.
+		// after the core API server gets the DELETE call, it will set the deletion timestamp
+		// as we expect, so we should proceed to remove the deletion timestamp & update as usual
+		// (below), so that the object is removed completely
+		if err := delete(t.cl, key, t.resourcePlural, ns, name, http.StatusOK); err != nil {
+			glog.Errorf("hard-deleting %s (%s)", key, err)
+			return err
+		}
+	}
+
+	t.cl.Get().Namespace(ns).Name(name).Do()
+
+	if err := get(
+		t.cl,
+		key,
+		t.resourcePlural,
+		ns,
+		name,
+		out,
+		false,
+	); err != nil {
+		glog.Errorf("getting %s (%s)", key, err)
+		return err
+	}
+
+	if _, err := scmeta.RemoveFinalizer(out, v1alpha1.FinalizerServiceCatalog); err != nil {
+		glog.Errorf("removing finalizer from %#v (%s)", out, err)
+		return err
+	}
+	if err := put(t.cl, t.resourcePlural, ns, name, out, out); err != nil {
+		glog.Errorf("putting %s (%s)", key, err)
+		return err
+	}
+	return nil
+}
+
+// Watch begins watching the specified key. Events are decoded into API objects,
+// and any items selected by 'p' are sent down to returned watch.Interface.
+// resourceVersion may be used to specify what version to begin watching,
+// which should be the current resourceVersion, and no longer rv+1
+// (e.g. reconnecting without missing any updates).
+// If resource version is "0", this interface will get current object at given key
+// and send it in an "ADDED" event, before watch starts.
+func (t *store) Watch(
+	ctx context.Context,
+	key string,
+	resourceVersion string,
+	p storage.SelectionPredicate,
+) (watch.Interface, error) {
+	ns, name, err := t.decodeKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	req := t.cl.Get().Namespace(ns).Name(name).Param("resourceVersion", resourceVersion)
+	watchIface, err := req.Watch()
+	if err != nil {
+		glog.Errorf("initiating the raw watch (%s)", err)
+		return nil, err
+	}
+	return watchIface, nil
+}
+
+// WatchList begins watching the specified key's items. Items are decoded into API
+// objects and any item selected by 'p' are sent down to returned watch.Interface.
+// resourceVersion may be used to specify what version to begin watching,
+// which should be the current resourceVersion, and no longer rv+1
+// (e.g. reconnecting without missing any updates).
+// If resource version is "0", this interface will list current objects directory defined by key
+// and send them in "ADDED" events, before watch starts.
+func (t *store) WatchList(
+	ctx context.Context,
+	key string,
+	resourceVersion string,
+	p storage.SelectionPredicate,
+) (watch.Interface, error) {
+	ns, _, err := t.decodeKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	req := t.cl.Get().Namespace(ns).Param("resourceVersion", resourceVersion)
+	watchIface, err := req.Watch()
+	if err != nil {
+		glog.Errorf("initiating the raw watch (%s)", err)
+		return nil, err
+	}
+	return watchIface, nil
```

Comment:
In the original TPR code this result is wrapped into filter: https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/storage/tpr/storage_interface.go#L249
@arschles As far as I understand, this method is essentially a "map" function (as a functional programming term) for transforming a collection of type X type into a collection of type Y?
Not sure we actually need this.

Reply:
yes, the filter is essentially a map function